### PR TITLE
Fix postion overflow in hybrid bulk scorer

### PR DESCRIPTION
**Pull request by vibrantvarun**
### Description

This PR contains the fix of position overflow of docId in HybridBulkScorer.
Earlier the method getNextDocIdCandidate finds the maxDocId and saves in the variable nextDoc.
Lets consider the following example to understand better how the bug is reproduced.
 Match query docId = [30004]
 Knn query docId = [4295, 4586, 37007]

### 1st iteration
Consider in a segment the maximum docId number is 38000

Scoring starts:
Scoring 1st window where min =0 and max = 4096
The [advance method ](https://github.com/opensearch-project/neural-search/blob/3.3/src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java#L158)will move the docIdIterator to first docId of each queries respectively. The docIds array becomes [30004, 4295]

The [allDocIdUsed](https://github.com/opensearch-project/neural-search/blob/d22991e2dbf4c4794967dde2c0384f2459ecd203/src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java#L172C60-L173C1) becomes false both the docIds are greater than 4096. Therefore, we move to get [getNextDocIdCandidate](https://github.com/opensearch-project/neural-search/blob/3.3/src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java#L181C17-L181C38).
It returns the max of both the Id's in the docId array i.e. 30004

### 2nd iteration
Scoring 2nd window where min = 30004 and max  30004+Interval_size
The max value returned from the first Iteration will become min for the next iteration as per [CancellableBulkScorer](https://github.com/opensearch-project/OpenSearch/blob/3.3/server/src/main/java/org/opensearch/search/internal/CancellableBulkScorer.java#L66).
So we start considering docIds that are greater than 30004 and **completely skip knn query** docId's 4295 and 4586/

The HQ results of query phase will contain docId 30004 and 37007.

This results in missing docIds that could have high scores.

### Fix
1. nextDoc variable instead of being a single variable should be an array.
2. It stores nextDoc that has results for each query.
3. Take minimum of both the docId's.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
https://github.com/opensearch-project/neural-search/issues/1705

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/RRFNormalizationTechnique.java [hunk 1]
```diff
@@ -125,8 +124,9 @@ private void processTopDocsEntry(
         int topDocsIndex,
         TriConsumer<DocIdAtSearchShard, Float, Integer> scoreProcessor
     ) {
-        for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
-            float normalizedScore = calculateNormalizedScore(Arrays.asList(topDocs.scoreDocs).indexOf(scoreDoc));
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/RRFNormalizationTechnique.java:129]: This was n^2 complexity due to usage of indexOf

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/RRFNormalizationTechnique.java:129]: Let's do this in a separate PR, this makes things complicated if we need to troubleshoot issues. I understood this is only optimization

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java [hunk 2]
```diff
@@ -179,13 +179,19 @@ private boolean allDocIdsUsed(int[] docsIds, int max) {
     }
 
     private int getNextDocIdCandidate(final int[] docsIds) {
-        int nextDoc = -1;
-        for (int doc : docsIds) {
-            if (doc != DocIdSetIterator.NO_MORE_DOCS) {
-                nextDoc = Math.max(nextDoc, doc);
+        int[] nextDoc = new int[docsIds.length];
+        Arrays.fill(nextDoc, -1);
+        int minDocIdForNextIteration = Integer.MAX_VALUE;
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: initialize with max value may be not safe, isn't NO_MORE_DOCS a better option?

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java [hunk 3]
```diff
@@ -179,13 +179,19 @@ private boolean allDocIdsUsed(int[] docsIds, int max) {
...
+        for (int i = 0; i < nextDoc.length; i++) {
+            if (docIds[i] != DocIdSetIterator.NO_MORE_DOCS) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: should this read from `docsIds`. Same for the next line

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java [hunk 4]
```diff
@@ -179,13 +179,19 @@ private boolean allDocIdsUsed(int[] docsIds, int max) {
...
+                nextDoc[i] = Math.max(nextDoc[i], docIds[i]);
+                if (nextDoc[i] < minDocIdForNextIteration) {
+                    minDocIdForNextIteration = nextDoc[i];
+                }
             }
         }
-        return nextDoc == -1 ? DocIdSetIterator.NO_MORE_DOCS : nextDoc;
+
+        return minDocIdForNextIteration == -1 ? DocIdSetIterator.NO_MORE_DOCS : minDocIdForNextIteration;
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: is this scenario even possible: `minDocIdForNextIteration == -1`? I think this is dead code now

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java [hunk 5]
```diff
@@ -179,13 +179,19 @@ private boolean allDocIdsUsed(int[] docsIds, int max) {
     }
 
     private int getNextDocIdCandidate(final int[] docsIds) {
-        int nextDoc = -1;
-        for (int doc : docsIds) {
-            if (doc != DocIdSetIterator.NO_MORE_DOCS) {
-                nextDoc = Math.max(nextDoc, doc);
+        int[] nextDoc = new int[docsIds.length];
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: do we really need `nextDoc` array? It's filled with `-1`, then `Math.max(-1, docIds[i])` is always `docIds[i]` (since doc IDs are non-negative). And array is never used for anything else
I think following code is doing the same, but is simpler:
```
private int getNextDocIdCandidate(final int[] docsIds) {
    int minDocId = Integer.MAX_VALUE;
    for (int doc : docsIds) {
        if (doc != DocIdSetIterator.NO_MORE_DOCS && doc < minDocId) {
            minDocId = doc;
        }
    }
    return minDocId == Integer.MAX_VALUE ? DocIdSetIterator.NO_MORE_DOCS : minDocId;
}
```

File: src/test/java/org/opensearch/neuralsearch/query/HybridBulkScorerTests.java [hunk 6]
```diff
@@ -106,4 +106,107 @@ public void testScoreWithInvalidRange() throws IOException {
...
+        when(mockIterator2.docID()).thenReturn(-1);
+        when(mockIterator2.advance(0)).thenReturn(DocIdSetIterator.NO_MORE_DOCS);
+        when(mockIterator2.nextDoc()).thenReturn(DocIdSetIterator.NO_MORE_DOCS);
+
+        // Execute scoring
+        int result = bulkScorer.score(mockLeafCollector, null, 0, MAX_DOC);
+
+        // Should return NO_MORE_DOCS when all scorers are exhausted
+        assertEquals("Should return NO_MORE_DOCS when all scorers exhausted", DocIdSetIterator.NO_MORE_DOCS, result);
+    }
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/query/HybridBulkScorerTests.java:211]: we need more test cases, but that can be added in later PRs:
1. multi-window iteration test
test that multiple windows are processed correctly with advancing doc IDs:
2. window boundary edge cases
3. sparse scorer (don't confuse with sparse query)  test

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java (hunk 5 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: Ah Infact we can even get rid of this condition

doc != DocIdSetIterator.NO_MORE_DOCS if we initialize by DocIdSetIterator.NO_MORE_DOCS.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/RRFNormalizationTechnique.java (hunk 1 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/RRFNormalizationTechnique.java:129]: ack.

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java (hunk 2 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: yes. Integer.MaxValue and No_more_docs value is same

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java (hunk 4 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: Change the code itself.

File: src/test/java/org/opensearch/neuralsearch/query/HybridBulkScorerTests.java (hunk 6 above)
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/query/HybridBulkScorerTests.java:211]: ack.

File: src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java (hunk 3 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java]: Changed the code itself/

**Review by heemin32 — APPROVED** [review]
